drivers/ethernet/eth_stm32_hal.c

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright (c) 2017 Erwin Rol <erwin@erwinrol.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_DOMAIN "dev/eth_stm32_hal"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_ETHERNET_LEVEL
+#include <logging/sys_log.h>
+
+#include <kernel.h>
+#include <device.h>
+#include <misc/__assert.h>
+#include <misc/util.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <net/net_pkt.h>
+#include <net/net_if.h>
+#include <soc.h>
+#include <misc/printk.h>
+#include <clock_control.h>
+#include <clock_control/stm32_clock_control.h>
+
+#include <stm32f4xx_hal_eth.h>
+
+#include "eth_stm32_hal_priv.h"
+
+static ETH_DMADescTypeDef dma_rx_desc_tab[ETH_RXBUFNB] __aligned(4);
+static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] __aligned(4);
+static u8_t dma_rx_buffer[ETH_RXBUFNB][ETH_RX_BUF_SIZE] __aligned(4);
+static u8_t dma_tx_buffer[ETH_TXBUFNB][ETH_TX_BUF_SIZE] __aligned(4);
+
+static int eth_tx(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct device *dev;
+	struct eth_stm32_hal_dev_data *dev_data;
+	ETH_HandleTypeDef *heth;
+	u8_t *dma_buffer;
+	int res;
+	struct net_buf *frag;
+	u16_t total_len;
+	__IO ETH_DMADescTypeDef *dma_tx_desc;
+
+	__ASSERT_NO_MSG(iface != NULL);
+	__ASSERT_NO_MSG(pkt != NULL);
+	__ASSERT_NO_MSG(pkt->frags != NULL);
+
+	dev = net_if_get_device(iface);
+	dev_data = DEV_DATA(dev);
+
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG(dev_data != NULL);
+
+	heth = &dev_data->heth;
+
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
+
+	total_len = net_pkt_ll_reserve(pkt) + net_pkt_get_len(pkt);
+	if (total_len > ETH_TX_BUF_SIZE) {
+		SYS_LOG_ERR("PKT to big\n");
+		res = -EIO;
+		goto error;
+	}
+
+	dma_tx_desc = heth->TxDesc;
+	while ((dma_tx_desc->Status & ETH_DMATXDESC_OWN) != (u32_t)RESET) {
+		k_yield();
+	}
+
+	dma_buffer = (u8_t *)(dma_tx_desc->Buffer1Addr);
+
+	memcpy(dma_buffer, net_pkt_ll(pkt),
+		net_pkt_ll_reserve(pkt) + pkt->frags->len);
+	dma_buffer += net_pkt_ll_reserve(pkt) + pkt->frags->len;
+
+	frag = pkt->frags->frags;
+	while (frag) {
+		memcpy(dma_buffer, frag->data, frag->len);
+		dma_buffer += frag->len;
+		frag = frag->frags;
+	}
+
+	if (HAL_ETH_TransmitFrame(heth, total_len) != HAL_OK) {
+		SYS_LOG_ERR("HAL_ETH_TransmitFrame failed\n");
+		res = -EIO;
+		goto error;
+	}
+
+	res = 0;
+error:
+	/* When Transmit Underflow flag is set, clear it and issue a
+	 * Transmit Poll Demand to resume transmission.
+	 */
+	if ((heth->Instance->DMASR & ETH_DMASR_TUS) != (u32_t)RESET) {
+		/* Clear TUS ETHERNET DMA flag */
+		heth->Instance->DMASR = ETH_DMASR_TUS;
+		/* Resume DMA transmission*/
+		heth->Instance->DMATPDR = 0;
+	}
+
+	net_pkt_unref(pkt);
```

Comment:
The driver should only unref the pkt if it could send it properly. If there was an error, then do not unref the packet but return value < 0 so that caller can decide what to do with the packet in this case.